File: src/actions/bet.php
<?php
/**
 * Copyright Hermann Krumrey <hermann@krumreyh.com> 2017
 *
 * This file is part of bundesliga_tippspiel.
 *
 * bundesliga_tippspiel is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * bundesliga_tippspiel is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with bundesliga_tippspiel. If not, see <http://www.gnu.org/licenses/>.
 */

namespace bundesliga_tippspiel_actions;
require __DIR__ . '/../../vendor/autoload.php';
use bundesliga_tippspiel\Functions;
use cheetah\BetManager;
use cheetah\Match;
use cheetah\SeasonManager;
use welwitschi\Authenticator;
use ErrorException;

Functions::initializeSession();

/**
 * Places a bet
 */
function bet() {

	$auth = new Authenticator(Functions::getMysqli());
	$db = Functions::getMysqli();
	$seasonManager = new SeasonManager($db);

	if (!isset($_POST["matchday_referrer"]) ||
		(int)$_POST["matchday_referrer"] < 1 ||
		(int)$_POST["matchday_referrer"] > $seasonManager->getMaxMatchday()) {
		$_SESSION["message"] = [
			"type" => "danger",
			"title" => "@{BET_FAIL_INVALID_MATCHDAY_TITLE}",
			"body" => "@{BET_FAIL_INVALID_MATCHDAY_BODY}"
		];
		header("Location: ../bets.php");

	} elseif (isset($_SESSION["user_id"])) {
		$user = $auth->getUserFromId($_SESSION["user_id"]);
		if ($user !== null) {

			$betManager = new BetManager($db);

			$matchDay = (int)$_POST["matchday_referrer"];
			$matches = Match::getAllForMatchday($db, $matchDay);

			$errors = false;

			foreach ($matches as $match) {
				$homeScore = $_POST[$match->homeTeam->id];
				$awayScore = $_POST[$match->awayTeam->id];

				if ($homeScore === "" || $awayScore === "") {
					continue;  // Skip unentered fields
				}

				$homeScore = (int)$homeScore;
				$awayScore = (int)$awayScore;

				if ($homeScore < 1000 && $homeScore >= 0 &&
					$awayScore < 1000 && $awayScore >= 0) {
					if (!$betManager->placeBetWithLoginSession(
						$user, $match, $homeScore, $awayScore)
					) {
						$errors = true;
					}
				} else {
					$errors = true;
				}
			}

			if ($errors) {
				$_SESSION["message"] = [
					"type" => "warning",
					"title" => "@{BET_SUCCESS_WITH_ERRORS_TITLE}",
					"body" => "@{BET_SUCCESS_WITH_ERRORS_BODY}"
				];
			} else {
				$_SESSION["message"] = [
					"type" => "success",
					"title" => "@{BET_SUCCESS_TITLE}",
					"body" => "@{BET_SUCCESS_BODY}"
				];
			}
			header("Location: ../bets.php?matchday=" . (string)$matchDay);

		} else {
			header("Location: ../index.php");
		}
	} else {
		header("Location: ../index.php");
	}
}

// Make ErrorException catch everything
set_error_handler(function($errno, $errstr, $errfile, $errline) {
	throw new ErrorException($errstr, $errno, 0, $errfile, $errline);
});
try {
	bet();
} catch (ErrorException $e) {
	echo "Oops... Something broke on our end, sorry!";
	throw $e;
}

File: bundesliga_tippspiel/db/user_generated/Bet.py
"""LICENSE
Copyright 2017 Hermann Krumrey <hermann@krumreyh.com>

This file is part of bundesliga-tippspiel.

bundesliga-tippspiel is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

bundesliga-tippspiel is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with bundesliga-tippspiel.  If not, see <http://www.gnu.org/licenses/>.
LICENSE"""

from typing import Any, List, Optional
from jerrycan.base import db
from jerrycan.db.ModelMixin import ModelMixin
from jerrycan.db.User import User
from bundesliga_tippspiel.db.match_data.Match import Match


class Bet(ModelMixin, db.Model):
    """
    Model that describes the 'bets' SQL table
    """

    MAX_POINTS: int = 15
    POSSIBLE_POINTS: List[int] = [0, 3, 7, 10, 12, 15]

    def __init__(self, *args, **kwargs):
        """
        Initializes the Model
        :param args: The constructor arguments
        :param kwargs: The constructor keyword arguments
        """
        super().__init__(*args, **kwargs)

    __tablename__ = "bets"
    __table_args__ = (
        db.ForeignKeyConstraint(
            ("home_team_abbreviation", "away_team_abbreviation",
             "league", "season", "matchday"),
            (Match.home_team_abbreviation, Match.away_team_abbreviation,
             Match.league, Match.season, Match.matchday)
        ),
    )

    league: str = db.Column(db.String(255), primary_key=True)
    season: int = db.Column(db.Integer, primary_key=True)
    matchday: int = db.Column(db.Integer, primary_key=True)
    home_team_abbreviation: str = db.Column(db.String(3), primary_key=True)
    away_team_abbreviation: str = db.Column(db.String(3), primary_key=True)
    user_id: int = db.Column(
        db.Integer,
        db.ForeignKey("users.id"),
        primary_key=True
    )

    home_score: int = db.Column(db.Integer, nullable=False)
    away_score: int = db.Column(db.Integer, nullable=False)
    points: int = db.Column(db.Integer, nullable=True)

    user: User = db.relationship(
        "User", backref=db.backref("bets", cascade="all, delete")
    )
    match: Match = db.relationship("Match", overlaps="bets")

    def __repr__(self) -> str:
        """
        :return: A string with which the object may be generated
        """
        params = ""

        for key, val in self.__json__().items():
            if key == "points":
                continue
            params += "{}={}, ".format(key, repr(val))
        params = params.rsplit(",", 1)[0]

        return "{}({})".format(self.__class__.__name__, params)

    def __eq__(self, other: Any) -> bool:
        """
        Checks the model object for equality with another object
        :param other: The other object
        :return: True if the objects are equal, False otherwise
        """
        if isinstance(other, Bet):
            return self.user_id == other.user_id \
                   and self.home_team_abbreviation == \
                   other.home_team_abbreviation \
                   and self.away_team_abbreviation == \
                   other.away_team_abbreviation \
                   and self.home_score == other.home_score \
                   and self.away_score == other.away_score
        else:
            return False  # pragma: no cover

    def evaluate(self) -> Optional[int]:
        """
        Evaluates the current points score on this bet
        :return: The calculated points (or None if the math hasn't started yet)
        """
        if not self.match.has_started:
            return None

        points = 0
        bet_diff = self.home_score - self.away_score
        match_diff = \
            self.match.home_current_score - self.match.away_current_score

        if bet_diff == match_diff:  # Correct goal difference
            points += 5

        if bet_diff * match_diff > 0:  # Correct winner
            points += 7
        elif bet_diff == 0 and match_diff == 0:  # Draw
            points += 7

        if self.home_score == self.match.home_current_score \
                or self.away_score == self.match.away_current_score:
            points += 3

        return points
